src/librustc_resolve/lib.rs

```diff
@@ -766,25 +766,30 @@ impl<'a, 'tcx> Visitor<'tcx> for Resolver<'a> {
         // to following type parameters, as the Substs can only
         // provide previous type parameters as they're built.
         let mut default_ban_rib = Rib::new(ForwardTyParamBanRibKind);
-        default_ban_rib.bindings.extend(generics.ty_params.iter()
+        default_ban_rib.bindings.extend(generics.params.iter()
+            .flat_map(|p| if let GenericParam::Type(ref tp) = *p { Some(tp) } else { None })
```

Comment:
Shouldn't this use `.ty_params()`? Although, in the long term, we should be moving named lifetime resolution to this pass.

Reply:
I only added `lifetimes()` and `ty_params()` to hir, not libsyntax, given that the former is part of the rustc crate that already uses `conservative_impl_trait` and the latter doesn't currently use any nightly features.

Reply:
Ah that's fine then.